KAFKA-6948 Change comparison to avoid overflow inconsistencies

thisthat opened the pull request:
Change timestamp comparison following what the Java documentation recommends to help preventing such errors.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-6028: Improve the quota throttle communication (KIP-219)

This implements KIP-219, where a broker returns a response with throttle time on
quota violation immediately after processing the corresponding request.  After
the response is sent out, the broker will keep the channel muted until the
throttle time is over. Also, on receiving a response with throttle time, client
will block outgoing communication to the broker for the specified throttle time.

See PR 4830, 5064 and 5094 for all the review history

Author: Jon Lee <jonlee@jonlee-ld1.linkedin.biz>

Reviewers: Jun Rao <junrao@gmail.com>, Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk>,  Dong Lin <lindong28@gmail.com>

Closes #5064 from jonlee2/kip-219

Commit:
Change comparison to avoid overflow inconsistencies

Change comparison to avoid overflow inconsistencies

Fix interface with close method

Commit:
Change comparison to avoid overflow inconsistencies

Commit:
Change comparison to avoid overflow inconsistencies

ijuma committed:
MINOR: Use reflection for signal handler and do not enable it for IBM JDK (#5047)

The Signal classes are not available in the compile classpath
if --release is used so we use reflection as a workaround.
As part of that moved the code to Java and added a simple
unit test.

Also disabled the signal handler if the IBM JDK is being used
due to KAFKA-6918.

Manually tested shutdown via ctrl+c and verified that
the message is printed.

radai-rosenblatt committed:
Improve kafka client sensor registration performance by lazily calculating JMX attributes

When any metric (e.g. per-partition metric) is created or deleted,
registerMBean() is called which in turn calls getMBeanInfo().getClassName().
However, KafkaMbean.getMBeanInfo() instantiates an array of all sensors even
though we only need the class name. This costs a lot of CPU to register
sensors when consumer with large partition assignment starts. For example, it
takes 5 minutes to start a consumer with 35k partitions. This patch reduces the
consumer startup time seconds.

Author: radai-rosenblatt <radai.rosenblatt@gmail.com>

Reviewers: Satish Duggana <satish.duggana@gmail.com>, Dong Lin <lindong28@gmail.com>

Closes #5011 from radai-rosenblatt/fun-with-jmx

junrao committed:
KAFKA-6937: In-sync replica delayed during fetch if replica throttle is exceeded (#5074)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Dong Lin <lindong28@gmail.com>, Ben Stopford <benstopford@gmail.com>

darionyaphet committed:
KAFKA-6930: Convert byte array to string in KafkaZkClient debug log (#5061)

Reviewers: Ismael Juma <ismael@juma.me.uk>

maytals committed:
Minor: Fixed ConsumerOffset#path (#5060)

consumer offset path in zookeeper should be /consumers/${group}/offsets/${topic}/${partition} instead of /consumers/${group}/offset/${topic}/${partition}. Added `s` to the word `offset`.

Reviewers: Ismael Juma <ismael@juma.me.uk>, Manikumar Reddy O <manikumar.reddy@gmail.com>, Jun Rao <junrao@gmail.com>

hachikuji committed:
MINOR: AdminClient should respect retry backoff

AdminClient should backoff when retrying a Call. Fixed and added a unit test

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Dong Lin <lindong28@gmail.com>

Closes #5077 from hachikuji/admin-client-retry-backoff

omkreddy committed:
KAFKA-2951; Add a test to verify produce, consume with ACLs for topic/group wildcard resources (#5054)

rondagostino committed:
KAFKA-6562: OAuth Authentication via SASL/OAUTHBEARER (KIP-255) (#4994)

This KIP adds the following functionality related to SASL/OAUTHBEARER:

1) Allow clients (both brokers when SASL/OAUTHBEARER is the inter-broker protocol as well as non-broker clients) to flexibly retrieve an access token from an OAuth 2 authorization server based on the declaration of a custom login CallbackHandler implementation and have that access token transparently and automatically transmitted to a broker for authentication.

2) Allow brokers to flexibly validate provided access tokens when a client establishes a connection based on the declaration of a custom SASL Server CallbackHandler implementation.

3) Provide implementations of the above retrieval and validation features based on an unsecured JSON Web Token that function out-of-the-box with minimal configuration required (i.e. implementations of the two types of callback handlers mentioned above will be used by default with no need to explicitly declare them).

4) Allow clients (both brokers when SASL/OAUTHBEARER is the inter-broker protocol as well as non-broker clients) to transparently retrieve a new access token in the background before the existing access token expires in case the client has to open new connections.

chia7712 committed:
MINOR: Replace unused variables by underscore (#5003)

And remove one unused expression.

Reviewers: Ismael Juma <ismael@juma.me.uk>

vvcephei committed:
KAFKA-5697; Implement new consumer poll API from KIP-266 (#4855)

Add the new stricter-timeout version of `poll` proposed in KIP-266.

The pre-existing variant `poll(long timeout)` would block indefinitely for metadata
updates if they were needed, then it would issue a fetch and poll for `timeout` ms 
for new records. The initial indefinite metadata block caused applications to become
stuck when the brokers became unavailable. The existence of the timeout parameter
made the indefinite block especially unintuitive.

This PR adds `poll(Duration timeout)` with the semantics:
1. iff a metadata update is needed:
    1. send (asynchronous) metadata requests
    2. poll for metadata responses (counts against timeout)
        - if no response within timeout, **return an empty collection immediately**
2. if there is fetch data available, **return it immediately**
3. if there is no fetch request in flight, send fetch requests
4. poll for fetch responses (counts against timeout)
    - if no response within timeout, **return an empty collection** (leaving async fetch request for the next poll)
    - if we get a response, **return the response**

The old method, `poll(long timeout)` is deprecated, but we do not change its semantics, so it remains:
1. iff a metadata update is needed:
    1. send (asynchronous) metadata requests
    2. poll for metadata responses *indefinitely until we get it*
2. if there is fetch data available, **return it immediately**
3. if there is no fetch request in flight, send fetch requests
4. poll for fetch responses (counts against timeout)
    - if no response within timeout, **return an empty collection** (leaving async fetch request for the next poll)
    - if we get a response, **return the response**

One notable usage is prohibited by the new `poll`: previously, you could call `poll(0)` to block for metadata updates, for example to initialize the client, supposedly without fetching records. Note, though, that this behavior is not according to any contract, and there is no guarantee that `poll(0)` won't return records the first time it's called. Therefore, it has always been unsafe to ignore the response.

hachikuji commented:
Thanks for the PR. Which java documentation are you referring to?

ijuma committed:
MINOR: Remove MaxPermSize from gradle.properties

No longer needed since we dropped support for Java 7.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Dong Lin <lindong28@gmail.com>

Closes #5083 from ijuma/remove-max-perm-size

omkreddy committed:
MINOR: Update consumer javadoc for invalid operations on unassigned partitions  (#5005)

Document cases where  `IllegalStateException` is raised when attempting an invalid operation on an unassigned partition. Also change `position()` to raise `IllegalStateException` when called on an unassigned partition for consistency.

Commit:
MINOR: Remove findbugs exclusion matching removed old producer (#5090)

KAFKA-6921 removed deprecated scala producer. This pull request removes the now unnecessary findbugs exclusion that matched one of the affected classes.

rajinisivaram committed:
KAFKA-6916; Refresh metadata in admin client if broker connection fails (#5050)

Refresh metadata if broker connection fails so that new calls are sent only to nodes that are alive and requests to controller are sent to the new controller if controller changes due to broker failure. Also reassign calls that could not be sent.

Reviewers: Dong Lin <lindong28@gmail.com>, Jason Gustafson <jason@confluent.io>

C0urante committed:
KAFKA-5540: Deprecate internal converter configs (KIP-174)

Implementation of [KIP-174](https://cwiki.apache.org/confluence/display/KAFKA/KIP-174+-+Deprecate+and+remove+internal+converter+configs+in+WorkerConfig)

Configuration properties 'internal.key.converter' and 'internal.value.converter'
are deprecated, and default to org.apache.kafka.connect.json.JsonConverter.

Warnings are logged if values are specified for either, or if properties that
appear to configure instances of internal converters (i.e., ones prefixed with
either 'internal.key.converter.' or 'internal.value.converter.') are given.

The property 'schemas.enable' is also defaulted to false for internal
JsonConverter instances (both for keys and values) if it isn't specified.

Documentation and code have also been updated with deprecation notices and
annotations, respectively.

Unit tests have been updated in `PluginsTest` to account for the new defaults for `schemas.enable` for internal key/value converters, and to ensure that (for the time being), internal key/value converters are still configurable despite being deprecated.

Author: Chris Egerton <chrise@confluent.io>
Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #4693 from C0urante/kafka-5540

bbejeck committed:
MINOR: reduce commit time on test (#5095)

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <guozhang@confluent.io>

rhauch committed:
KAFKA-6913: Add Connect converters and header converters for short, integer, long, float, and double (KIP-305)

*[KIP-305](https://cwiki.apache.org/confluence/display/KAFKA/KIP-305%3A+Add+Connect+primitive+number+converters) has been approved.*

Added converters and header converters for the primitive number types for which Kafka already had serializers and deserializers. All extend a common base class, `NumberConverter`, that encapsulates most of the shared functionality. Unit tests were added to check the basic functionality.

These classes are not used by any other Connect code, and must be explicitly used in Connect workers and connectors.

Author: Randall Hauch <rhauch@gmail.com>

Reviewers: Arjun Satish <wicknicks@users.noreply.github.com>, Magesh Nandakumar <magesh.n.kumar@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5034 from rhauch/kafka-6913

mageshn committed:
KAFKA-6776: ConnectRestExtension Interfaces & Rest integration (KIP-285)

This PR provides the implementation for KIP-285 and also a reference implementation for authenticating BasicAuth credentials using JAAS LoginModule

Author: Magesh Nandakumar <magesh.n.kumar@gmail.com>

Reviewers: Randall Hauch <rhauch@gmail.com>, Arjun Satish <wicknicks@users.noreply.github.com>, Konstantine Karantasis <konstantine@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #4931 from mageshn/KIP-285

Commit:
KAFKA-6617; Improve controller performance by batching reassignment znode write operation

KafkaController currently writes reassignment znode once for every partition that has been successfully reassigned. This is unnecessary and controller should be able to update reassignment znode once to remove all partitions that have been reassigned from the reassignment znode.

Author: Dong Lin <dolin@linkedin.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #4659 from lindong28/KAFKA-6617

ConcurrencyPractitioner committed:
KAFKA-6608; Add timeout parameter to blocking consumer calls [KIP-266] (#5014)

This patch implements the consumer timeout APIs from KIP-266 (everything except `poll()`, which was done separately).

Reviewers:  John Roesler <john@confluent.io>, Jason Gustafson <jason@confluent.io>

thisthat commented:
@hachikuji thank you for the answer and sorry for the delay! 
The documentation that I am referring to the following: [link](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--)

joan38 committed:
Update doc of curried joins and aggregates (#5053)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Dark0096 committed:
Remove duplicate code which is invoked twice (#5039)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Guozhang Wang <wangguoz@gmail.com>

fhussonnois committed:
KAFKA-6957 make InternalTopologyBuilder accessible from AbstractStream subclasses (#5085)

Currently, the AbstractStream class defines a copy-constructor that allow to extend KStream and KTable APIs with new methods without impacting the public interface.

However adding new processor or/and store to the topology is made throught the internalTopologyBuilder that is not accessible from AbstractStream subclasses defined outside of the package (package visibility).

Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <guozhang@confluent.io>

wicknicks committed:
KAFKA-6738: Implement error handling for source and sink tasks (KIP-298)

This PR implements the features described in this KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-298%3A+Error+Handling+in+Connect

This PR changes the Connect framework to allow it to automatically deal with errors encountered while processing records in a Connector. The following behavior changes are introduced here:

**Retry on Failure**: Retry the failed operation a configurable number of times, with backoff between each retry.
**Task Tolerance Limits**: Tolerate a configurable number of failures in a task.

We also add the following ways to report errors, along with sufficient context to simplify the debugging process:

**Log Error Context**: The error information along with processing context is logged along with standard application logs.
**Dead Letter Queue**: Produce the original message into a Kafka topic (applicable only to sink connectors).

New **metrics** which will monitor the number of failures, and the behavior of the response handler are added.

The changes proposed here **are backward compatible**. The current behavior in Connect is to kill the task on the first error in any stage. This will remain the default behavior if the connector does not override any of the new configurations which are provided as part of this feature.

Testing: added multiple unit tests to test the retry and tolerance logic.

Author: Arjun Satish <arjun@confluent.io>
Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Magesh Nandakumar <magesh.n.kumar@gmail.com>, Randall Hauch <rhauch@gmail.com>, Konstantine Karantasis <konstantine@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5065 from wicknicks/KAFKA-6378

guozhangwang committed:
KAFKA-4936: Add dynamic routing in Streams (#5018)

implements KIP-303

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Matthias J. Sax <matthias@confluent.io>

rayokota committed:
KAFKA-6886: Externalize secrets from Connect configs (KIP-297)

This commit allows secrets in Connect configs to be externalized and replaced with variable references of the form `${provider:[path:]key}`, where the "path" is optional.

There are 2 main additions to `org.apache.kafka.common.config`: a `ConfigProvider` and a `ConfigTransformer`.  The `ConfigProvider` is an interface that allows key-value pairs to be provided by an external source for a given "path".  An a TTL can be associated with the key-value pairs returned from the path.  The `ConfigTransformer` will use instances of `ConfigProvider` to replace variable references in a set of configuration values.

In the Connect framework, `ConfigProvider` classes can be specified in the worker config, and then variable references can be used in the connector config.  In addition, the herder can be configured to restart connectors (or not) based on the TTL returned from a `ConfigProvider`.  The main class that performs restarts and transformations is `WorkerConfigTransformer`.

Finally, a `configs()` method has been added to both `SourceTaskContext` and `SinkTaskContext`.  This allows connectors to get configs with variables replaced by the latest values from instances of `ConfigProvider`.

Most of the other changes in the Connect framework are threading various objects through classes to enable the above functionality.

Author: Robert Yokota <rayokota@gmail.com>
Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5068 from rayokota/KAFKA-6886-connect-secrets

hachikuji committed:
MINOR: Follow-up improvements for KIP-266 (#5084)

This patch contains a few follow-up improvements/cleanup for KIP-266:

- Add upgrade notes
- Add missing `commitSync(Duration)` API
- Improve timeout messages and fix some naming inconsistencies
- Various small cleanups

Reviewers: John Roesler <john@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
KAFKA-6054: Add 'version probing' to Kafka Streams rebalance (#4636)

implements KIP-268

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

dhruvilshah3 committed:
KAFKA-6927; Chunked down-conversion to prevent out of memory errors on broker [KIP-283] (#4871)

Implementation for lazy down-conversion in a chunked manner for efficient memory usage during down-conversion. This pull request is mainly to get initial feedback on the direction of the patch. The patch includes all the main components from KIP-283.

Reviewers: Jason Gustafson <jason@confluent.io>

viktorsomogyi commented on clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java:
```diff
@@ -65,7 +65,7 @@ public static boolean awaitReady(KafkaClient client, Node node, Time time, long
 
         long attemptStartTime = time.milliseconds();
-        while (!client.isReady(node, attemptStartTime) && attemptStartTime < expiryTime) {
+        while (!client.isReady(node, attemptStartTime) && attemptStartTime - expiryTime < 0) {
```
Would it make sense to create a utility function for this given it's used in 9 places?

viktorsomogyi reviewed (commented):
According to the javadocs of `System.nanoTime()` and `Syatem.currentTimeMillis()` overflows can happen only with the first one. How is it possible with the second?

joan38 committed:
KAFKA-6936: Implicit materialized for aggregate, count and reduce (#5066)

In #4919 we propagate the SerDes for each of these aggregation operators.

As @guozhangwang mentioned in that PR:

```
reduce: inherit the key and value serdes from the parent XXImpl class.
count: inherit the key serdes, enforce setting the Serdes.Long() for value serdes.
aggregate: inherit the key serdes, do not set for value serdes internally.
```

Although it's all good for reduce and count, it is quiet unsafe to have aggregate without Materialized given. In fact I don't see why we would not give a Materialized for the aggregate since the result type will always be different (otherwise use reduce) and also the value Serde is simply not propagated.

This has been discussed previously in a broader PR before but I believe for aggregate we could pass implicitly a Materialized the same way we pass a Joined, just to avoid the stupid case. Then if the user wants to specialize, he can give his own Materialized.

Reviewers: Debasish Ghosh <dghosh@acm.org>, Guozhang Wang <guozhang@confluent.io>

emmanuelharel committed:
MINOR:Fix table outer join test (#5099)

huxi-2b committed:
KAFKA-6973: Validate topic config message.timestamp.type (#5106)

Specifying an invalid config (i.e. something other than `CreateTime` or
`LogAppendTime`) via `TopicCommand` would previously cause the
broker to fail on start-up.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

bbejeck committed:
MINOR: Use thread name and task for sensor name (#5111)

Changes to keep the operation name as is and make the sensor name unique.

Reviewers: John Roesler <john@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

mimaison committed:
KAFKA-6760: Fix response logging in the Controller (#4834)

- Override toString in LeaderAndIsrResponse and StopReplicaResponse
- Add unit tests

Reviewers: Ismael Juma <ismael@juma.me.uk>

vahidhashemian committed:
KAFKA-6955: Use Java AdminClient in DeleteRecordsCommand (#5088)

- Removed internal kafka.admin.AdminClient.deleteRecordsBefore since it's
no longer used.
- Removed redundant tests and rewrote non redundant ones to use the Java
AdminClient.

Reviewers: Viktor Somogyi <viktor.somogyi@cloudera.com>, Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

rayokota committed:
MINOR: Fix JMX serialization by reverting lazy computation of JMX attributes (#5114)

This reverts commit c9ec292135f26d1d70a (#5011). That
commit introduces an anonymous inner class which retains a
reference to the non-serializable outer class `KafkaMbean`
breaking Serialization. This means that reading JMX metrics
via JConsole or JmxTool no longer works since RMI relies
on Java Serialization.

Reviewers: Jason Gustafson <jason@confluent.io>, Dong Lin <lindong28@gmail.com>, Ismael Juma <ismael@juma.me.uk>

rootex- committed:
KAFKA-6743; ConsumerPerformance fails to consume all messages [KIP-281]  (#4818)

This patch implements KIP-281, which adds a configurable timeout to the consumer performance tool with a default value of 10 seconds. The old timeout was hard-coded as 1 second. Additionally, this patch adds a warning message when the tool exits after a timeout rather than returning silently.

Reviewers: Dhruvil Shah <dhruvil@confluent.io>, Jason Gustafson <jason@confluent.io>

hachikuji committed:
HOTFIX: Fix compilation error in ConsumerPerformance (#5127)

Reviewers: Ismael Juma <ismael@juma.me.uk>

rajinisivaram committed:
KAFKA-6912; Add test for authorization with custom principal types

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Dong Lin <lindong28@gmail.com>

Closes #5030 from rajinisivaram/MINOR-group-test

vahidhashemian committed:
MINOR: Remove usages of JavaConversions and fix some typos (#5115)

Reviewers: Ismael Juma <ismael@juma.me.uk>

omkreddy committed:
KAFKA-5523: Remove ReplayLogProducer tool (#5092)

Reviewers: Ismael Juma <ismael@juma.me.uk>

guozhangwang commented:
I'm okay with this change, and also I think it is better as a general coding style suggestion to use the proposed pattern whenever possible.

@thisthat could you rebase?

thisthat commented:
@guozhangwang sure! Should I create an utility function as @viktorsomogyi suggested or just the rebase is sufficient?

thisthat commented:
@viktorsomogyi the documentation presents the error only for the nano method call. However, the overflow can happen also with the millis call, e.g. KAFKA-4290.

hachikuji committed:
MINOR: Fix bug in AdminClient node reassignment following connection failure (#5112)

We added logic to reassign nodes in callToSend after a connection failure, but we do not handle the case when there is no node currently available to reassign the request to. This can happen when using MetadataUpdateNodeIdProvider if all of the known nodes are blacked out awaiting the retry backoff. To fix this, we need to ensure that the call is added to pendingCalls if a new node cannot be found.

guozhangwang commented:
I think it is okay to just fix all the places: even if we introduce the util function, we cannot enforce future contributors to leverage on it. I think it is more practical to ask reviewers be more keen to such coding styles moving forward.

guozhangwang committed:
MINOR: Remove deprecated KafkaStreams constructors in docs (#5118)

Reviewers: Bill Bejeck <bill@confluent.io>, Matthias J. Sax <matthias@confluent.io>

ppatierno committed:
KAFKA-5919; Adding checks on "version" field for tools using it

Adding checks on "version" field for tools using it.
This is a new version of the closed PR #3887 (to see for more comments and related discussion).

Author: Paolo Patierno <ppatierno@live.com>

Reviewers: Dong Lin <lindong28@gmail.com>

Closes #5126 from ppatierno/kafka-5919-update

vvcephei committed:
KAFKA-6813: return to double-counting for count topology names (#5075)

#4919 unintentionally changed the topology naming scheme. This change returns to the prior scheme.

Reviewers: Bill Bejeck <bill@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

jeqo committed:
KAFKA-6938: Add documentation for accessing Headers on Kafka Streams Processor API (#5128)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Commit:
Merge branch 'trunk' of https://github.com/thisthat/kafka into trunk

rajinisivaram committed:
KAFKA-3665: Enable TLS hostname verification by default (KIP-294) (#4956)

Make HTTPS the default ssl.endpoint.identification.algorithm.

Reviewers: Ismael Juma <ismael@juma.me.uk>

mimaison committed:
KAFKA-6750: Add listener name to authentication context (KIP-282) (#4829)

PrincipalBuilder implementations can now take the listener into account
when creating the Principal. This is especially interesting in deployments
where inter-broker traffic is on a different listener than client traffic or
when the same protocol is used by multiple listeners.

The change in itself is mostly "plumbing" as the listener name needs to be
passed from ChannelBuilders all the way down to all classes implementing
AuthenticationContext.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Ismael Juma <ismael@juma.me.uk>

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

vvcephei committed:
MINOR: Add regression tests for KTable mapValues and filter (#5134)

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

guozhangwang committed:
MINOR: Add missing configs for resilience settings

Reviewers: Matthias J. Sax <matthias@confluent.io>, Bill Bejeck <bill@confluent.io>

bbejeck committed:
KAFKA-6704: InvalidStateStoreException from IQ when StreamThread closes store (#4801)

While using an iterator from IQ, it's possible to get an InvalidStateStoreException if the StreamThread closes the store during a range query.

Added a unit test to SegmentIteratorTest for this condition.

Reviewers: John Roesler <john@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

rajinisivaram committed:
MINOR: Rename package `internal` to `internals` for consistency (#5137)

ConcurrencyPractitioner committed:
[KAFKA-6730] Simplify State Store Recovery (#5013)

Reviewer: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <guozhang@confluent.io>, Bill Bejeck <bill@confluent.io>

wicknicks committed:
KAFKA-6981: Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes (KIP-298)

Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes, and refactor the tests and classes to use these new properties.

Testing: Unit tests and running the connect-standalone script with a file sink connector.

Author: Arjun Satish <arjun@confluent.io>
Author: Randall Hauch <rhauch@gmail.com>

Reviewers: Konstantine Karantasis <konstantine@confluent.io>, Magesh Nandakumar <magesh.n.kumar@gmail.com>, Robert Yokota <rayokota@gmail.com>, Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5125 from wicknicks/KAFKA-6981

mjsax committed:
MINOR: docs should point to latest version (#5132)

Reviewers: Guozhang Wang <wangguoz@gmail.com>

mageshn committed:
KAFKA-6997: Exclude test-sources.jar when $INCLUDE_TEST_JARS is FALSE

Exclude test-sources.jar when $INCLUDE_TEST_JARS is FALSE

Author: Magesh Nandakumar <magesh.n.kumar@gmail.com>

Reviewers: Konstantine Karantasis <konstantine@confluent.io>, Robert Yokota <rayokota@gmail.com>, Arjun Satish <arjun@confluent.io>, Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5139 from mageshn/KAFKA-6997

KoenDG committed:
MINOR: Fix for the location of the trogdor.sh executable file in the documentation. (#5040)

Reviewers: Colin Patrick McCabe <colin@cmccabe.xyz>, Ismael Juma <ismael@juma.me.uk>

ppatierno committed:
KAFKA-5588: Remove deprecated --new-consumer tools option (#5097)

Reviewers: Viktor Somogyi <viktorsomogyi@gmail.com>, Vahid Hashemian <vahidhashemian@us.ibm.com>, Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

edoardocomar committed:
KAFKA-6726: Fine Grained ACL for CreateTopics (KIP-277) (#4795)

- CreateTopicsRequest now requires Create auth on Topic resource
or Create on Cluster resource.
- AclCommand --producer option adjusted
- Existing unit and Integration tests adjusted accordingly and
new tests added.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

big-andy-coates committed:
KAFKA-6841: Support Prefixed ACLs (KIP-290) (#5117)

Reviewers: Colin Patrick McCabe <colin@cmccabe.xyz>, Jun Rao <junrao@gmail.com>

Co-authored-by: Piyush Vijay <pvijay@apple.com>
Co-authored-by: Andy Coates <big-andy-coates@users.noreply.github.com>

rayokota committed:
MINOR: Use service loader for ConfigProvider impls (KIP-297)

This is a small change to use the Java ServiceLoader to load ConfigProvider plugins.  It uses code added by mageshn for Connect Rest Extensions.

Author: Robert Yokota <rayokota@gmail.com>

Reviewers: Magesh Nandakumar <magesh.n.kumar@gmail.com>, Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5141 from rayokota/service-loader-for-config-plugins

rajinisivaram committed:
MINOR: Bump version to 2.1.0-SNAPSHOT (#5153)

dongjinleekr committed:
KAFKA-6993: Fix defective documentations for KStream/KTable methods (#5136)

* KAFKA-6993: Fix defective documentations for KStream/KTable methods

1. Fix the documentation of following methods, e.g., making more detailed description for the overloaded methods:

	- KStream#join
	- KStream#leftJoin
	- KStream#outerJoin
	- KTable#filter
	- KTable#filterNot
	- KTable#mapValues
	- KTable#transformValues
	- KTable#join
	- KTable#leftJoin
	- KTable#outerJoin

2. (trivial) with possible new type -> with possibly new type.

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <guozhang@confluent.io>, Bill Bejeck <bill@confluent.io>

mjsax committed:
KAFKA-6967: TopologyTestDriver does not allow pre-populating state stores that have change logging (#5096)

Reviewers: Guozhang Wang <guozhang@confluent.io>, James Cheng <jylcheng@yahoo.com>, Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>

mjsax committed:
MINOR: need to update system test version after version bump (#5156)

bbejeck committed:
KAFKA-6935: Add config for allowing optional optimization (#5071)

Adding configuration to StreamsConfig allowing for making topology optimization optional.

Added unit tests are verifying default values, setting correct value and failure on invalid values.

Reviewers: John Roesler <john@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

jadireddi committed:
KAFKA-6538: Changes to enhance  ByteStore exceptions thrown from RocksDBStore with more human readable info (#5103)

Reviewers: Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mageshn committed:
KAFKA-6991: Fix ServiceLoader issue with PluginClassLoader (KIP-285)

Fix ServiceLoader issue with PluginClassLoader and add basic-auth-extension packaging & classpath

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

Author: Magesh Nandakumar <magesh.n.kumar@gmail.com>

Reviewers: Konstantine Karantasis <konstantine@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5135 from mageshn/KAFKA-6991

Commit:
Fix interface with close method

ijuma committed:
MINOR: Upgrade Gradle to 4.8 and bug fix updates for other deps (#5148)

In addition to Gradle, updated snappy, owasp-dependency-check,
apache directory service api.

Gradle 4.8 fixes a fatal issue when building with Java 11, but
full support is coming in 4.9 or later.

thisthat commented:
@guozhangwang the rebase should be fine now

guozhangwang commented:
@thisthat Could you rebase? Seems like a lot of recent commits were reverted in the diff file.

wicknicks committed:
KAFKA-7001: Rename errors.allowed.max property in Connect to errors.tolerance (KIP-298)

Signed-off-by: Arjun Satish <arjunconfluent.io>

Author: Arjun Satish <arjun@confluent.io>

Reviewers: Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5146 from wicknicks/KAFKA-7001

wicknicks committed:
KAFKA-7002: Add a config property for DLQ topic's replication factor (KIP-298)

Currently, the replication factor is hardcoded to a value of 3. This means that we cannot use a DLQ in any cluster setup with less than three brokers. It is better to have the user specify this value if the default value does meet the requirements.

Testing: A unit test is added.

Signed-off-by: Arjun Satish <arjunconfluent.io>

Author: Arjun Satish <arjun@confluent.io>

Reviewers: Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #5145 from wicknicks/KAFKA-7002

big-andy-coates committed:
KAFKA-7011 - Remove ResourceNameType field from Java Resource class. (#5160)

The initial PR for KIP-290 #5117 added a `ResourceNameType` field to the Java and Scala `Resource` classes to introduce the concept of Prefixed ACLS.  This does not make a lot of sense as these classes are meant to represent cluster resources, which would not have a concept of 'name type'. This work has not been released yet, so we have time to change it.

This PR looks to refactor the code to remove the name type field from the Java `Resource` class. (The Scala one will age out once KIP-290 is done, and removing it would involve changes to the `Authorizer` interface, so this class was not touched).

This is achieved by replacing the use of `Resource` with `ResourcePattern` and `ResourceFilter` with `ResourceFilterPattern`.  A `ResourcePattern` is a combination of resource type, name and name type, where each field needs to be defined. A `ResourcePatternFilter` is used to select patterns during describe and delete operations.

The adminClient uses `AclBinding` and `AclBindingFilter`. These types have been switched over to use the new pattern types.

The AclCommands class, used by Kafka-acls.sh, has been converted to use the new pattern types.

The result is that the original `Resource` and `ResourceFilter` classes are not really used anywhere, except deprecated methods. However, the `Resource` class will be used if/when KIP-50 is done.

Reviewers: Colin Patrick McCabe <colin@cmccabe.xyz>, Jun Rao <junrao@gmail.com>

mjsax committed:
MINOR: remove duplicate map in StoreChangelogReader (#5143)

Reviewers: Guozhang Wang <guozhang@confluent.io>, John Roesler <john@confluent.io>, Bill Bejeck <bill@confluent.io>

Commit:
Merge branch 'trunk' of https://github.com/thisthat/kafka into trunk

thisthat commented:
@guozhangwang rebased!

guozhangwang commented:
Hmm.. it still contains 20K+ diffs, if the branch is a bit messed you could consider creating a new branch out of current trunk.

thisthat commented:
@guozhangwang agree! I will create a new PR and close this one 😉
